File: src/main/python/Sorting/quick_sort.py
class QuickSort:

    @staticmethod
    def __partition(arr: list[int], start: int, end: int) -> int:
        pivot = arr[start]
        i = start + 1
        j = end

        while True:

            while i <= end and arr[i] < pivot:
                i += 1

            while j > start and arr[j] > pivot:
                j -= 1

            if i >= j:
                break

            arr[i], arr[j] = arr[j], arr[i]
            i += 1
            j -= 1

        arr[start], arr[j] = arr[j], arr[start]
        return j

    @staticmethod
    def __recursive_sort(arr: list[int], start: int, end: int) -> None:
        if start >= end:
            return

        pivot = QuickSort.__partition(arr, start, end)

        QuickSort.__recursive_sort(arr, start, pivot - 1)
        QuickSort.__recursive_sort(arr, pivot + 1, end)

    @staticmethod
    def recursive_sort(arr: list[int]) -> None:
        if len(arr) <= 1:
            return

        QuickSort.__recursive_sort(arr, 0, len(arr) - 1)

    @staticmethod
    def iterative_sort(arr: list[int]) -> None:
        if len(arr) <= 1:
            return

        stack = [(0, len(arr) - 1)]

        while stack:
            start, end = stack.pop()

            if start >= end:
                continue

            pivot = QuickSort.__partition(arr, start, end)

            stack.append((start, pivot - 1))
            stack.append((pivot + 1, end))

File: src/main/java/Sorting/QuickSort.java
package main.java.Sorting;

import java.util.Stack;

public class QuickSort {

    private static void swap(int[] arr, int x, int y) {
        int temp = arr[x];
        arr[x] = arr[y];
        arr[y] = temp;
    }

    private static int partition(int[] arr, int start, int end) {

        int pivot = arr[start];
        int i = start + 1;
        int j = end;

        while (true) {

            while (i <= end && arr[i] < pivot)
                i++;

            while (j > start && arr[j] > pivot)
                j--;

            if (i >= j)
                break;

            swap(arr, i, j);
            i++;
            j--;
        }

        swap(arr, start, j);
        return j;
    }

    private static void recursiveSort(int[] arr, int start, int end) {

        if (start >= end)
            return;

        int p = partition(arr, start, end);

        recursiveSort(arr, start, p - 1);
        recursiveSort(arr, p + 1, end);
    }

    public static void recursiveSort(int[] arr) {
        recursiveSort(arr, 0, arr.length - 1);
    }

    public static void iterativeSort(int[] arr) {

        Stack<int[]> stack = new Stack<>();
        stack.push(new int[]{0, arr.length - 1});

        while (!stack.isEmpty()) {

            int[] range = stack.pop();

            int start = range[0];
            int end = range[1];

            if (start >= end)
                continue;

            int p = partition(arr, start, end);

            stack.push(new int[]{start, p - 1});
            stack.push(new int[]{p + 1, end});
        }
    }
}